### ISSUE TITLE
Can't connect to events from JSON in F#

### ISSUE THREAD

**Issue by rmunn** [picoe/Eto#392]
I'm trying to use Eto from F#, and I'd like to use JSON rather than XAML to define my GUI since I'm writing it by hand. But the following doesn't seem to work:

Relevant JSON snippet from Panel1.jeto: `{ "$type": "Button", "ID": "FirstBtn", "Text": "Click Me", "Click": "FirstClicked" }`
My F# code:
```
type EventDelegate = delegate of EventArgs -> unit

type Panel1 () as this =
    inherit Form ()

    do
        JsonReader.Load(this, "Panel1.jeto")

    member this.FirstClicked e =
        ignore(MessageBox.Show(this, "You clicked the button!"))
```

But when I run this, I get:

```
ArgumentException was unhandled.

An unhandled exception of type 'System.ArgumentException' occurred in Newtonsoft.Json.dll

Additional information: Cannot bind to the target method because its signature or security transparency is not compatible with that of the delegate type.
```

I've been unable to figure out what function signature the system is looking for. I would have expected it to be a method function that takes a single EventArgs parameter, as in my code above. But turning `FirstClicked` into a static method that takes two parameters (an object and an EventArgs) has not helped either.

The example code that I've been able to find for using Eto in F# (i.e., the code that gets created in the default Eto app template in Visual Studio) uses an approach like the following:

```
let clickMe = new Command(MenuText = "Click Me!", ToolBarText = "Click Me!")
clickMe.Executed.Add(fun e -> ignore(MessageBox.Show(this, "I was clicked!")))
```

This would be a perfectly fine approach if I could use it. But to use this approach, I'd have to be able to get the Button object that was created by the JSON serialization. I've given it an ID, so you'd think I could just look it up by ID -- but I can't find a "GetChildById()" method *anywhere* in the Eto API. I've found Control.FindParent() which takes an ID, but there seems to be no FindChild() method. And to get the FirstButton object out of my JSON-serialized Panel, I'd need to be able to call panel.FindChild("FirstButton").

So I'm stuck. I've been completely unable to attach an event to an object coming out of a JSON file, and I've found no way to do so. Is there some obvious way that I'm missing? Or am I going to have to go back to creating panel objects "by hand", since the JSON serializer is useless without the ability to attach event handlers to GUI objects?

**Comment by cwensley** [comment]
To connect to an event using json (or xaml), you need to define your event handler with the correct parameter and return types.. All of Eto's event delegate requires the first parameter to be an Object, and the second an EventArgs (or subclass of EventArgs).  The `FirstClicked` method you defined roughly translates to this in C#:

```c#
void FirstClicked<T>(T e)
{
  MessageBox.Show(this, "You clicked the button!");
}
```

To do this correctly, you need to define the two event parameters and explicitly define their types:

```fsharp
member this.FirstClicked (sender:Object, e:EventArgs) =
        ignore(MessageBox.Show(this, "You clicked the button!"))
```

which translates to this in C#:

```c#
void FirstClicked(object sender, EventArgs e)
{
  MessageBox.Show(this, "You clicked the button!");
}
```
I will be working on adding some initial documentation for using json/xaml with F# when I have some time, though help in that regard would also be much appreciated!

**Comment by rmunn** [comment]
I'd be happy to help with F# documentation if I can find the time, though don't expect anything huge: this isn't my main responsibility at work right now so I only get bits and pieces of time on this. Also, I'm just learning F# myself so my examples may not be the best. I do have some experience with functional languages (mostly Clojure, with a tiny bit of Haskell and Elm) so I'm not a complete newbie, but my F# code examples might not be the best F# you've ever seen. :smiley:

**Comment by rmunn** [comment]
Yep, declaring my event handlers explicitly like `member this.FirstClicked (sender:Object, e:EventArgs)` worked. I'd tried declaring them with two parameters before, but I hadn't tried explicitly annotating their types.

... Actually, I just tried declaring the event handler as `member this.FirstClicked (sender, e)` and it also worked. So apparently the type signatures don't have to be explicit and can be inferred. And apparently I didn't try declaring that the handler takes two parameters as a tuple; I had only tried declaring that it takes two parameters in normal F# style (`member this.FirstClicked sender e` ) which isn't the right signature for event handlers.

Thanks: declaring the handler's signature as a tuple worked -- and I can also drop the type signatures and it will still work.
